include/RAJA/policy/atomic_builtin.hpp

```diff
@@ -219,17 +177,52 @@ template <typename T, typename OPER>
 RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomic_CAS_oper(T volatile *acc,
                                                       OPER &&oper)
 {
-  BuiltinAtomicCAS<sizeof(T)> cas;
-  return cas(acc, std::forward<OPER>(oper), [](T const &) { return false; });
+  static_assert(sizeof(T) == 4 || sizeof(T) == 8,
+                "builtin atomic cas assumes 4 or 8 byte targets");
+
+#ifdef RAJA_COMPILER_MSVC
+#pragma warning( disable : 4244 )  // Force msvc to not emit conversion warning
+#endif
+  T old = builtin_atomic_load(acc);
+  T assumed;
+
+  do {
+    assumed = old;
+    old = builtin_atomic_CAS(acc, assumed, oper(assumed));
+  } while (assumed != old);
+
+  return old;
+#ifdef RAJA_COMPILER_MSVC
+#pragma warning( default : 4244 )  // Reenable warning
+#endif
 }
 
 template <typename T, typename OPER, typename ShortCircuit>
 RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomic_CAS_oper_sc(T volatile *acc,
                                                          OPER &&oper,
                                                          ShortCircuit const &sc)
 {
-  BuiltinAtomicCAS<sizeof(T)> cas;
-  return cas(acc, std::forward<OPER>(oper), sc);
+  static_assert(sizeof(T) == 4 || sizeof(T) == 8,
+                "builtin atomic cas assumes 4 or 8 byte targets");
+
+#ifdef RAJA_COMPILER_MSVC
+#pragma warning( disable : 4244 )  // Force msvc to not emit conversion warning
+#endif
+  T old = builtin_atomic_load(acc);
+
+  if (!sc(old)) {
```

Comment:
Would it be cleaner to have an early return?

Reply:
Done.